### PR TITLE
Windows Compatibility Fixes #1

### PR DESCRIPTION
**Pull request by yeastplume**
Meta Issue: #2525 

After this PR, Grin should compile, run and sync on windows (with remaining problems, see #2525 for instructions and remaining issues). In particular, this:

* Updates `croaring-rs` from 0.3 to 0.3.8
* Remove `daemonize` and `grin server stop|start` commands
* Update inclusion of `cursive` to be platform conditional, change terminal backend for Windows to pancurses
* Updates to PMMR (particularly file rewind) to allow for explicit dropping of file handles. Required due to stricter locking on windows
* Modifications to store and wallet tests (which also need to explicitly drop file handles in many cases)
* Temporarily reduce initial LMDB file storage on windows (more complete solution needed, see #2525)
* Add functions to explicitly drop underlying file handles in TxHashSet. Not doing so was preventing fast-sync archives from unzipping.

These changes definitely need to be reviewed and tested to ensure they don't break anything on linux

### CODE REVIEW

File: store/src/pmmr.rs [hunk 1]
```diff
@@ -53,8 +53,8 @@ pub const PMMR_FILES: [&str; 4] = [
 pub struct PMMRBackend<T: PMMRable> {
 	data_dir: PathBuf,
 	prunable: bool,
-	hash_file: DataFile<Hash>,
-	data_file: DataFile<T::E>,
+	hash_file: Option<DataFile<Hash>>,
+	data_file: Option<DataFile<T::E>>,
```
Comment (yeastplume) [inline store/src/pmmr.rs]: Options so they can be dropped before renaming operations

File: store/src/types.rs [hunk 2]
```diff
@@ -193,21 +193,37 @@ impl AppendOnlyFile {
...
+					.create(true)
+					.write(true)
+					.open(&self.path)?;
+				file.set_len(self.buffer_start as u64)?;
+			}
+			let file = OpenOptions::new()
+				.read(true)
+				.create(true)
+				.append(true)
+				.open(&self.path)?;
```
Comment (yeastplume) [inline store/src/types.rs]: File permissions for `set_len` on windows need to be as set here or it fails. Hence the drop and reopen with original permissions after the `set_len`.

File: store/src/lmdb.rs [hunk 3]
```diff
@@ -73,11 +73,19 @@ pub fn new_named_env(path: String, name: String, max_readers: Option<u32>) -> lm
...
-		});
+
+	#[cfg(not(target_os = "windows"))]
+	env_builder.set_mapsize(5_368_709_120).unwrap_or_else(|e| {
+		panic!("Unable to allocate LMDB space: {:?}", e);
+	});
+	//TODO: This is temporary to support (beta) windows support
+	//Windows allocates the entire file at once, so this needs to
+	//be changed to allocate as little as possible and increase as needed
+	#[cfg(target_os = "windows")]
```
Comment (i1skn) [inline store/src/lmdb.rs]: Can we add `cfg!(target_os = "ios")` here as well?

Comment (yeastplume) [reply store/src/lmdb.rs]: I'd rather just keep this PR to windows issues for now, and this is just temporary. The real fix shouldn't involve OS specific code, but it should allocate the file minimally and increase it as necessary.

**Comment by yeastplume** [comment]
@ignopeverell Thanks, cleaned that up and encapsulated the file rename/replace. However, needs slightly more review given that I had to add functionality to drop underlying PMMR files from the txhashset during fast sync archive unzip. Marked above.

File: chain/src/chain.rs [hunk 4]
```diff
@@ -857,6 +857,14 @@ impl Chain {
 		}
 
 		let header = self.get_block_header(&h)?;
+
+		{
+			let mut txhashset_ref = self.txhashset.write();
+			// Drop file handles in underlying txhashset
+			txhashset_ref.release_backend_files();
```
Comment (yeastplume) [inline chain/src/chain.rs]: Seems to work, would just like a bit of reassurance that this isn't going to have any odd side effects.

File: store/src/lmdb.rs (hunk 3 above)
Comment (i1skn) [reply store/src/lmdb.rs]: ok, got it! let me know if I can help somehow with this upcoming fix!

**Comment by yeastplume** [comment]
Also, removed Daemonize altogether, think I agree it's far easier to use external commands to run something as a service in most cases

File: chain/src/chain.rs (hunk 4 above)
Comment (ignopeverell) [reply chain/src/chain.rs]: Not that I can think of, seems fair.

**Pull request merged**